Unable to load custom rules

According to the readme custom rules can be used, much like the maven plugin: (although the use of "probably" doesn't inspire confidence!)

> package your rule into a custom jar, add a dependency to this jar (probably with test scope) and declare the rule in the config

But when I try this, with my custom jar as a `testCompile` dependency, I see a cryptic failure which turns out to be an inability to load classes. It looks like the rules are being load via the context class loader using shared code, but it doesn't look like any of the project dependencies are in the plugin's classpath under Gradle.

```
:checkRules (Thread[Execution worker for ':',5,main]) completed. Took 1.069 secs.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':checkRules'.
> com.example.product.archunit.TestClassNamingRule

* Try:
Run with --debug option to get more log output. Run with --scan to get full insights.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':checkRules'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:200)
        at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:263)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:198)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:179)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:109)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:62)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$call$2(DefaultBuildOperationExecutor.java:76)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.callWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:54)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:76)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:41)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:372)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:359)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:352)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:338)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: com.societegenerale.commons.plugin.utils.ReflectionException: com.example.product.archunit.TestClassNamingRule
        at com.societegenerale.commons.plugin.utils.ReflectionException.wrap(ReflectionException.java:12)
        at com.societegenerale.commons.plugin.utils.ReflectionUtils.loadClassWithContextClassLoader(ReflectionUtils.java:13)
        at com.societegenerale.commons.plugin.service.RuleInvokerService.invokePreConfiguredRule(RuleInvokerService.java:74)
        at com.societegenerale.commons.plugin.service.RuleInvokerService.invokeRules(RuleInvokerService.java:59)
        at com.societegenerale.commons.plugin.gradle.ArchUnitRulesTask.checkRules(ArchUnitRulesTask.java:48)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:104)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.doExecute(StandardTaskAction.java:58)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:51)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:29)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$3.run(ExecuteActionsTaskExecuter.java:555)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:56)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$run$1(DefaultBuildOperationExecutor.java:71)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.runWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:45)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:71)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:540)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:523)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.access$300(ExecuteActionsTaskExecuter.java:108)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution.executeWithPreviousOutputFiles(ExecuteActionsTaskExecuter.java:271)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution.execute(ExecuteActionsTaskExecuter.java:260)
        at org.gradle.internal.execution.steps.ExecuteStep.lambda$execute$1(ExecuteStep.java:34)
        at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:34)
        at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:26)
        at org.gradle.internal.execution.steps.CleanupOutputsStep.execute(CleanupOutputsStep.java:67)
        at org.gradle.internal.execution.steps.CleanupOutputsStep.execute(CleanupOutputsStep.java:36)
        at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:49)
        at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:34)
        at org.gradle.internal.execution.steps.CancelExecutionStep.execute(CancelExecutionStep.java:43)
        at org.gradle.internal.execution.steps.TimeoutStep.executeWithoutTimeout(TimeoutStep.java:73)
        at org.gradle.internal.execution.steps.TimeoutStep.execute(TimeoutStep.java:54)
        at org.gradle.internal.execution.steps.CreateOutputsStep.execute(CreateOutputsStep.java:44)
        at org.gradle.internal.execution.steps.SnapshotOutputsStep.execute(SnapshotOutputsStep.java:54)
        at org.gradle.internal.execution.steps.SnapshotOutputsStep.execute(SnapshotOutputsStep.java:38)
        at org.gradle.internal.execution.steps.BroadcastChangingOutputsStep.execute(BroadcastChangingOutputsStep.java:42)
        at org.gradle.internal.execution.steps.CacheStep.executeWithoutCache(CacheStep.java:159)
        at org.gradle.internal.execution.steps.CacheStep.execute(CacheStep.java:72)
        at org.gradle.internal.execution.steps.CacheStep.execute(CacheStep.java:43)
        at org.gradle.internal.execution.steps.StoreExecutionStateStep.execute(StoreExecutionStateStep.java:44)
        at org.gradle.internal.execution.steps.StoreExecutionStateStep.execute(StoreExecutionStateStep.java:33)
        at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:38)
        at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:24)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.executeBecause(SkipUpToDateStep.java:92)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.lambda$execute$0(SkipUpToDateStep.java:85)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:55)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:39)
        at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:76)
        at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:37)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:36)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:26)
        at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:94)
        at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:49)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:79)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:53)
        at org.gradle.internal.execution.steps.ValidateStep.execute(ValidateStep.java:74)
        at org.gradle.internal.execution.steps.SkipEmptyWorkStep.lambda$execute$2(SkipEmptyWorkStep.java:78)
        at org.gradle.internal.execution.steps.SkipEmptyWorkStep.execute(SkipEmptyWorkStep.java:78)
        at org.gradle.internal.execution.steps.SkipEmptyWorkStep.execute(SkipEmptyWorkStep.java:34)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsStartedStep.execute(MarkSnapshottingInputsStartedStep.java:39)
        at org.gradle.internal.execution.steps.LoadExecutionStateStep.execute(LoadExecutionStateStep.java:40)
        at org.gradle.internal.execution.steps.LoadExecutionStateStep.execute(LoadExecutionStateStep.java:28)
        at org.gradle.internal.execution.impl.DefaultWorkExecutor.execute(DefaultWorkExecutor.java:33)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:187)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:179)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:109)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:62)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$call$2(DefaultBuildOperationExecutor.java:76)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.callWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:54)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:76)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:41)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:372)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:359)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:352)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:338)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: java.lang.ClassNotFoundException: com.example.product.archunit.TestClassNamingRule
        at com.societegenerale.commons.plugin.utils.ReflectionUtils.loadClassWithContextClassLoader(ReflectionUtils.java:11)
        ... 101 more
```

--- a/README.md
+++ b/README.md
@@ -72,7 +72,20 @@ To use the plugin, your `build.gradle` require these changes:
 ### Adding new rules
 
 All rules referenced in the configuration have to be available in the classpath. Therefore, you have 2 solutions : 
-- package your rule into a custom jar, add a dependency to this jar (probably with `test` scope) and declare the rule in the config
+- package your rule into a custom jar, add a dependency to this jar with the dedicated  configuration : 
+
+```
+dependencies {
+    ...
+    archUnitExtraLib('org.example:archunit-custom-rules:1.0.0')
+    ...
+}
+```
+
+**caveat** : when packaging your custom rules, if you're using `arch-unit-build-plugin-core`, make sure it's the same version that is used in this plugin. If it's not the same version and there are breaking changes, it may not work : the plugin may take some classes from "your" jar instead of its own jar, and it may fail if some classes are missing or if some method signatures have changed   
+
+
+
 - Propose your rule through a PullRequest to [Arch-Unit-Build-Plugin-Core](https://github.com/societe-generale/arch-unit-build-plugin-core) : if it's accepted, it will be part of the next release and usable by everyone. 
 
 ### Releasing a new version of the plugin

--- a/build.gradle
+++ b/build.gradle
@@ -19,7 +19,7 @@ repositories {
 
 dependencies {
     testImplementation group: 'junit', name: 'junit', version: '4.12'
-    implementation 'com.societegenerale.commons:arch-unit-build-plugin-core:2.9.4'
+    implementation 'com.societegenerale.commons:arch-unit-build-plugin-core:2.9.5'
 }
 
 task sourceJar(type: Jar) {

--- a/gradle.properties
+++ b/gradle.properties
@@ -1,2 +1,2 @@
-version=2.9.4-SNAPSHOT
+version=2.9.5-SNAPSHOT
 group=com.societegenerale.commons

--- a/src/main/java/com/societegenerale/commons/plugin/gradle/ArchUnitGradleConfig.java
+++ b/src/main/java/com/societegenerale/commons/plugin/gradle/ArchUnitGradleConfig.java
@@ -41,6 +41,14 @@ public class ArchUnitGradleConfig {
         return new ApplyOn(packageName,scope);
     }
 
+    public List<String> getPreConfiguredRules() {
+        return preConfiguredRules;
+    }
+
+    public List<ConfigurableRule> getConfigurableRules() {
+        return configurableRules;
+    }
+
     public List<String> check(String... checks){ return Arrays.asList(checks); }
 
     public Rules getRules() {

--- a/src/main/java/com/societegenerale/commons/plugin/gradle/ArchUnitGradlePlugin.java
+++ b/src/main/java/com/societegenerale/commons/plugin/gradle/ArchUnitGradlePlugin.java
@@ -1,22 +1,24 @@
 package com.societegenerale.commons.plugin.gradle;
 
 import org.gradle.api.*;
+import org.gradle.api.artifacts.Configuration;
 
 
 public class ArchUnitGradlePlugin implements Plugin<Project> {
 
     public void apply(Project project) {
 
-        ArchUnitGradleConfig archUnitGradleConfig = project.getExtensions().create("archUnit", ArchUnitGradleConfig.class, project);
+        Configuration conf = project.getConfigurations().create("archUnitExtraLib");
+        ArchUnitGradleConfig archUnitGradleConfig = project.getExtensions()
+            .create("archUnit", ArchUnitGradleConfig.class, project);
 
-        ArchUnitRulesTask archUnitTask=project.getTasks().create("checkRules", ArchUnitRulesTask.class, archUnitGradleConfig);
-
-        archUnitTask.setProjectBuildDir(project.getBuildDir());
+        ArchUnitRulesTask archUnitTask = project.getTasks().create("checkRules", ArchUnitRulesTask.class);
         archUnitTask.setGroup("verification");
+        archUnitTask.getClasspath().from(conf);
+        archUnitTask.getArchUnitGradleConfig().convention(archUnitGradleConfig);
 
-        final Task checkTask = findExistingTaskOrFailOtherwise("check",project);
-
-        final Task testTask = findExistingTaskOrFailOtherwise("test",project);
+        final Task checkTask = findExistingTaskOrFailOtherwise("check", project);
+        final Task testTask = findExistingTaskOrFailOtherwise("test", project);
 
         checkTask.dependsOn(archUnitTask);
         archUnitTask.mustRunAfter(testTask);

--- a/src/main/java/com/societegenerale/commons/plugin/gradle/ArchUnitRulesTask.java
+++ b/src/main/java/com/societegenerale/commons/plugin/gradle/ArchUnitRulesTask.java
@@ -8,7 +8,12 @@ import java.io.IOException;
 import org.apache.commons.lang3.StringUtils;
 import org.gradle.api.DefaultTask;
 import org.gradle.api.GradleException;
+import org.gradle.api.file.ConfigurableFileCollection;
+import org.gradle.api.tasks.Classpath;
+import org.gradle.api.tasks.InputFiles;
 import org.gradle.api.tasks.TaskAction;
+import org.gradle.workers.WorkQueue;
+import org.gradle.workers.WorkerExecutor;
 import org.slf4j.LoggerFactory;
 
 import javax.inject.Inject;
@@ -30,7 +35,7 @@ public class ArchUnitRulesTask extends DefaultTask {
     }
 
     @TaskAction
-    public void checkRules() throws InstantiationException, IllegalAccessException, InvocationTargetException {
+    public void checkRules() throws NoSuchMethodException, InstantiationException, IllegalAccessException, InvocationTargetException {
 
         if (archUnitGradleConfig.isSkip()) {
             logger.warn("Rule checking has been skipped!");
@@ -43,28 +48,27 @@ public class ArchUnitRulesTask extends DefaultTask {
             throw new GradleException("Arch unit Gradle Plugin should have at least one preconfigured/configurable rule!");
         }
 
-        String ruleFailureMessage = "";
 
-        try {
 
-            RuleInvokerService ruleInvokerService = new RuleInvokerService(new GradleLogAdapter(LoggerFactory.getLogger(RuleInvokerService.class)),
-                                                                            new GradleScopePathProvider(archUnitGradleConfig),
-                                                                            archUnitGradleConfig.getExcludedPaths(),
-                                                                            projectBuildDir.getCanonicalPath());
+        RuleInvokerService ruleInvokerService = new RuleInvokerService(new GradleLogAdapter(LoggerFactory.getLogger(RuleInvokerService.class)),
+                                                                        new GradleScopePathProvider(archUnitGradleConfig),
+                                                                        archUnitGradleConfig.getExcludedPaths(),
+                                                                        projectBuildDir.getCanonicalPath());
 
-            ruleFailureMessage = ruleInvokerService.invokeRules(rules);
-        } catch (IOException e) {
-            throw new RuntimeException("problem with the build directory",e);
-        }
+        WorkQueue queue = getWorkerExecutor().classLoaderIsolation(spec -> spec.getClasspath().from(getClasspath()));
+        queue.submit(WorkerAction.class, params -> {
+                params.getRuleInvokerService().set(ruleInvokerService);
+                params.getRules().set(rules);
+        });
+
+/*
+        ruleFailureMessage = ruleInvokerService.invokeRules(rules);
 
         if (!StringUtils.isEmpty(ruleFailureMessage)) {
             throw new GradleException(PREFIX_ARCH_VIOLATION_MESSAGE + " \n" + ruleFailureMessage);
         }
-    }
 
-    public void setProjectBuildDir(File projectBuildDir) {
-        this.projectBuildDir = projectBuildDir;
+ */
     }
-
 
 }

--- a/src/main/java/com/societegenerale/commons/plugin/gradle/ArchUnitRulesTask.java
+++ b/src/main/java/com/societegenerale/commons/plugin/gradle/ArchUnitRulesTask.java
@@ -2,6 +2,7 @@ package com.societegenerale.commons.plugin.gradle;
 
 import com.societegenerale.commons.plugin.Log;
 import com.societegenerale.commons.plugin.model.Rules;
+import java.util.stream.Collectors;
 import org.gradle.api.DefaultTask;
 import org.gradle.api.GradleException;
 import org.gradle.api.file.ConfigurableFileCollection;
@@ -54,7 +55,22 @@ public abstract class ArchUnitRulesTask extends DefaultTask {
             params.getTestClassesPath().set(archUnitGradleConfig.getBuildPath() + "/classes/java/test");
             params.getProjectBuildDirPath().set(archUnitGradleConfig.getBuildPath());
             params.getPreConfiguredRules().addAll(archUnitGradleConfig.getPreConfiguredRules());
-            params.getConfigurableRules().addAll(archUnitGradleConfig.getConfigurableRules());
+
+            params.getConfigurableRules().addAll(archUnitGradleConfig.getConfigurableRules().stream()
+                .map(configurableRule -> {
+                    WorkerActionParams.ApplyOn applyOn = new WorkerActionParams.ApplyOn();
+                    applyOn.packageName = configurableRule.getApplyOn().getPackageName();
+                    applyOn.scope = configurableRule.getApplyOn().getScope();
+
+                    WorkerActionParams.ConfigurableRule ret = new WorkerActionParams.ConfigurableRule();
+                    ret.rule = configurableRule.getRule();
+                    ret.applyOn = applyOn;
+                    ret.checks.addAll(configurableRule.getChecks());
+                    ret.skip = configurableRule.isSkip();
+
+                    return ret;
+                })
+                .collect(Collectors.toList()));
         });
 
     }

--- a/src/main/java/com/societegenerale/commons/plugin/gradle/ArchUnitRulesTask.java
+++ b/src/main/java/com/societegenerale/commons/plugin/gradle/ArchUnitRulesTask.java
@@ -2,14 +2,12 @@ package com.societegenerale.commons.plugin.gradle;
 
 import com.societegenerale.commons.plugin.Log;
 import com.societegenerale.commons.plugin.model.Rules;
-import com.societegenerale.commons.plugin.service.RuleInvokerService;
-import java.io.File;
-import java.io.IOException;
-import org.apache.commons.lang3.StringUtils;
 import org.gradle.api.DefaultTask;
 import org.gradle.api.GradleException;
 import org.gradle.api.file.ConfigurableFileCollection;
+import org.gradle.api.provider.Property;
 import org.gradle.api.tasks.Classpath;
+import org.gradle.api.tasks.Input;
 import org.gradle.api.tasks.InputFiles;
 import org.gradle.api.tasks.TaskAction;
 import org.gradle.workers.WorkQueue;
@@ -19,23 +17,25 @@ import org.slf4j.LoggerFactory;
 import javax.inject.Inject;
 import java.lang.reflect.InvocationTargetException;
 
-public class ArchUnitRulesTask extends DefaultTask {
+public abstract class ArchUnitRulesTask extends DefaultTask {
 
-    private ArchUnitGradleConfig archUnitGradleConfig;
+    @InputFiles
+    @Classpath
+    public abstract ConfigurableFileCollection getClasspath();
 
-    private File projectBuildDir;
-
-    private Log logger = new GradleLogAdapter(LoggerFactory.getLogger(ArchUnitRulesTask.class));
-
-    private static final String PREFIX_ARCH_VIOLATION_MESSAGE = "ArchUnit Gradle plugin reported architecture failures listed below :";
+    @Input
+    public abstract Property<ArchUnitGradleConfig> getArchUnitGradleConfig();
 
     @Inject
-    public ArchUnitRulesTask(ArchUnitGradleConfig archUnitGradleConfig) {
-         this.archUnitGradleConfig = archUnitGradleConfig;
-    }
+    protected abstract WorkerExecutor getWorkerExecutor();
+
+
+    private final Log logger = new GradleLogAdapter(LoggerFactory.getLogger(ArchUnitRulesTask.class));
 
     @TaskAction
-    public void checkRules() throws NoSuchMethodException, InstantiationException, IllegalAccessException, InvocationTargetException {
+    public void checkRules() {
+
+        ArchUnitGradleConfig archUnitGradleConfig = getArchUnitGradleConfig().get();
 
         if (archUnitGradleConfig.isSkip()) {
             logger.warn("Rule checking has been skipped!");
@@ -48,27 +48,15 @@ public class ArchUnitRulesTask extends DefaultTask {
             throw new GradleException("Arch unit Gradle Plugin should have at least one preconfigured/configurable rule!");
         }
 
-
-
-        RuleInvokerService ruleInvokerService = new RuleInvokerService(new GradleLogAdapter(LoggerFactory.getLogger(RuleInvokerService.class)),
-                                                                        new GradleScopePathProvider(archUnitGradleConfig),
-                                                                        archUnitGradleConfig.getExcludedPaths(),
-                                                                        projectBuildDir.getCanonicalPath());
-
         WorkQueue queue = getWorkerExecutor().classLoaderIsolation(spec -> spec.getClasspath().from(getClasspath()));
         queue.submit(WorkerAction.class, params -> {
-                params.getRuleInvokerService().set(ruleInvokerService);
-                params.getRules().set(rules);
+            params.getMainClassesPath().set(archUnitGradleConfig.getBuildPath() + "/classes/java/main");
+            params.getTestClassesPath().set(archUnitGradleConfig.getBuildPath() + "/classes/java/test");
+            params.getProjectBuildDirPath().set(archUnitGradleConfig.getBuildPath());
+            params.getPreConfiguredRules().addAll(archUnitGradleConfig.getPreConfiguredRules());
+            params.getConfigurableRules().addAll(archUnitGradleConfig.getConfigurableRules());
         });
 
-/*
-        ruleFailureMessage = ruleInvokerService.invokeRules(rules);
-
-        if (!StringUtils.isEmpty(ruleFailureMessage)) {
-            throw new GradleException(PREFIX_ARCH_VIOLATION_MESSAGE + " \n" + ruleFailureMessage);
-        }
-
- */
     }
 
 }

--- a/src/main/java/com/societegenerale/commons/plugin/gradle/GradleScopePathProvider.java
+++ b/src/main/java/com/societegenerale/commons/plugin/gradle/GradleScopePathProvider.java
@@ -5,21 +5,21 @@ import com.societegenerale.commons.plugin.service.ScopePathProvider;
 
 public class GradleScopePathProvider implements ScopePathProvider {
 
-    private ArchUnitGradleConfig archUnitGradleConfig;
+    private final RootClassFolder mainClassesPath;
+    private final RootClassFolder testClassesPath;
 
-    public GradleScopePathProvider(ArchUnitGradleConfig archUnitGradleConfig) {
-        this.archUnitGradleConfig=archUnitGradleConfig;
+    public GradleScopePathProvider(String mainPath, String testPath) {
+        mainClassesPath = new RootClassFolder(mainPath);
+        testClassesPath = new RootClassFolder(testPath);
     }
 
     @Override
     public RootClassFolder getMainClassesPath() {
-        return new RootClassFolder(
-                archUnitGradleConfig.getBuildPath() +"/classes/java/main");
+        return mainClassesPath;
     }
 
     @Override
     public RootClassFolder getTestClassesPath() {
-        return new RootClassFolder(
-                archUnitGradleConfig.getBuildPath() +"/classes/java/test");
+        return testClassesPath;
     }
 }

--- a/src/main/java/com/societegenerale/commons/plugin/gradle/WorkerAction.java
+++ b/src/main/java/com/societegenerale/commons/plugin/gradle/WorkerAction.java
@@ -1,10 +1,14 @@
 package com.societegenerale.commons.plugin.gradle;
 
+import com.societegenerale.commons.plugin.model.ApplyOn;
+import com.societegenerale.commons.plugin.model.ConfigurableRule;
 import java.lang.reflect.InvocationTargetException;
 import java.util.Collection;
 
 import com.societegenerale.commons.plugin.model.Rules;
 import com.societegenerale.commons.plugin.service.RuleInvokerService;
+import java.util.List;
+import java.util.stream.Collectors;
 import org.apache.commons.lang3.StringUtils;
 import org.gradle.api.GradleException;
 import org.gradle.workers.WorkAction;
@@ -30,7 +34,16 @@ public abstract class WorkerAction implements WorkAction<WorkerActionParams> {
 
         try {
 
-            Rules rules = new Rules(params.getPreConfiguredRules().get(), params.getConfigurableRules().get());
+            List<ConfigurableRule> configurableRules = params.getConfigurableRules().map(crules -> crules.stream()
+                    .map(crule -> new ConfigurableRule(
+                        crule.rule,
+                        new ApplyOn(crule.applyOn.packageName, crule.applyOn.scope),
+                        crule.checks,
+                        crule.skip))
+                    .collect(Collectors.toList()))
+                .get();
+            Rules rules = new Rules(params.getPreConfiguredRules().get(), configurableRules);
+
             ruleFailureMessage = ruleInvokerService.invokeRules(rules);
 
             if (!StringUtils.isEmpty(ruleFailureMessage)) {

--- a/src/main/java/com/societegenerale/commons/plugin/gradle/WorkerAction.java
+++ b/src/main/java/com/societegenerale/commons/plugin/gradle/WorkerAction.java
@@ -1,13 +1,8 @@
 package com.societegenerale.commons.plugin.gradle;
 
-import java.lang.reflect.Constructor;
 import java.lang.reflect.InvocationTargetException;
 import java.util.Collection;
-import java.util.List;
-import java.util.stream.Collectors;
 
-import com.societegenerale.commons.plugin.model.ApplyOn;
-import com.societegenerale.commons.plugin.model.ConfigurableRule;
 import com.societegenerale.commons.plugin.model.Rules;
 import com.societegenerale.commons.plugin.service.RuleInvokerService;
 import org.apache.commons.lang3.StringUtils;
@@ -26,22 +21,16 @@ public abstract class WorkerAction implements WorkAction<WorkerActionParams> {
         Collection<String> excludedPaths = params.getExcludedPaths().get();
 
         RuleInvokerService ruleInvokerService = new RuleInvokerService(
-                new GradleLogAdapter(LoggerFactory.getLogger(RuleInvokerService.class)),
-                new GradleScopePathProvider(params.getMainClassesPath().get(), params.getTestClassesPath().get()),
-                excludedPaths);
+            new GradleLogAdapter(LoggerFactory.getLogger(RuleInvokerService.class)),
+            new GradleScopePathProvider(params.getMainClassesPath().get(), params.getTestClassesPath().get()),
+            excludedPaths,
+            params.getProjectBuildDirPath().get());
 
         String ruleFailureMessage = "";
 
         try {
-            List<ConfigurableRule> configurableRules = params.getConfigurableRules().map(crules -> crules.stream()
-                            .map(crule -> new ConfigurableRule(
-                                    crule.rule,
-                                    new ApplyOn(crule.applyOn.packageName, crule.applyOn.scope),
-                                    crule.checks,
-                                    crule.skip))
-                            .collect(Collectors.toList()))
-                    .get();
-            Rules rules = new Rules(params.getPreConfiguredRules().get(), configurableRules);
+
+            Rules rules = new Rules(params.getPreConfiguredRules().get(), params.getConfigurableRules().get());
             ruleFailureMessage = ruleInvokerService.invokeRules(rules);
 
             if (!StringUtils.isEmpty(ruleFailureMessage)) {

--- a/src/main/java/com/societegenerale/commons/plugin/gradle/WorkerAction.java
+++ b/src/main/java/com/societegenerale/commons/plugin/gradle/WorkerAction.java
@@ -1,7 +1,14 @@
 package com.societegenerale.commons.plugin.gradle;
 
+import java.lang.reflect.Constructor;
 import java.lang.reflect.InvocationTargetException;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
 
+import com.societegenerale.commons.plugin.model.ApplyOn;
+import com.societegenerale.commons.plugin.model.ConfigurableRule;
+import com.societegenerale.commons.plugin.model.Rules;
 import com.societegenerale.commons.plugin.service.RuleInvokerService;
 import org.apache.commons.lang3.StringUtils;
 import org.gradle.api.GradleException;
@@ -14,18 +21,42 @@ public abstract class WorkerAction implements WorkAction<WorkerActionParams> {
 
     @Override
     public void execute() {
+        WorkerActionParams params = getParameters();
 
-        ArchUnitGradleConfig archUnitGradleConfig = getParameters().getArchUnitGradleConfig().get();
+        Collection<String> excludedPaths = params.getExcludedPaths().get();
 
-        RuleInvokerService ruleInvokerService = new RuleInvokerService(new GradleLogAdapter(LoggerFactory.getLogger(RuleInvokerService.class)),
-                new GradleScopePathProvider(archUnitGradleConfig),
-                archUnitGradleConfig.getExcludedPaths());
+        System.out.println("list of excludedPaths : "+excludedPaths);
+
+        Class ctClass = RuleInvokerService.class;
+
+        Constructor[] constructors = ctClass.getConstructors();
+        for (int i = 0; i < constructors.length; i++) {
+            System.out.println("constuctor: " + constructors[i]);
+        }
+
+        Constructor[] declaredConstructors = ctClass.getDeclaredConstructors();
+        for (int i = 0; i < declaredConstructors.length; i++) {
+            System.out.println("declared constructor: " + declaredConstructors[i]);
+        }
+
+        RuleInvokerService ruleInvokerService = new RuleInvokerService(
+                new GradleLogAdapter(LoggerFactory.getLogger(RuleInvokerService.class)),
+                new GradleScopePathProvider(params.getMainClassesPath().get(), params.getTestClassesPath().get()),
+                excludedPaths);
 
         String ruleFailureMessage = "";
 
         try {
-
-            ruleFailureMessage = ruleInvokerService.invokeRules(archUnitGradleConfig.getRules());
+            List<ConfigurableRule> configurableRules = params.getConfigurableRules().map(crules -> crules.stream()
+                            .map(crule -> new ConfigurableRule(
+                                    crule.rule,
+                                    new ApplyOn(crule.applyOn.packageName, crule.applyOn.scope),
+                                    crule.checks,
+                                    crule.skip))
+                            .collect(Collectors.toList()))
+                    .get();
+            Rules rules = new Rules(params.getPreConfiguredRules().get(), configurableRules);
+            ruleFailureMessage = ruleInvokerService.invokeRules(rules);
 
             if (!StringUtils.isEmpty(ruleFailureMessage)) {
                 throw new GradleException(PREFIX_ARCH_VIOLATION_MESSAGE + " \n" + ruleFailureMessage);

--- a/src/main/java/com/societegenerale/commons/plugin/gradle/WorkerAction.java
+++ b/src/main/java/com/societegenerale/commons/plugin/gradle/WorkerAction.java
@@ -25,20 +25,6 @@ public abstract class WorkerAction implements WorkAction<WorkerActionParams> {
 
         Collection<String> excludedPaths = params.getExcludedPaths().get();
 
-        System.out.println("list of excludedPaths : "+excludedPaths);
-
-        Class ctClass = RuleInvokerService.class;
-
-        Constructor[] constructors = ctClass.getConstructors();
-        for (int i = 0; i < constructors.length; i++) {
-            System.out.println("constuctor: " + constructors[i]);
-        }
-
-        Constructor[] declaredConstructors = ctClass.getDeclaredConstructors();
-        for (int i = 0; i < declaredConstructors.length; i++) {
-            System.out.println("declared constructor: " + declaredConstructors[i]);
-        }
-
         RuleInvokerService ruleInvokerService = new RuleInvokerService(
                 new GradleLogAdapter(LoggerFactory.getLogger(RuleInvokerService.class)),
                 new GradleScopePathProvider(params.getMainClassesPath().get(), params.getTestClassesPath().get()),

--- a/src/main/java/com/societegenerale/commons/plugin/gradle/WorkerAction.java
+++ b/src/main/java/com/societegenerale/commons/plugin/gradle/WorkerAction.java
@@ -1,0 +1,41 @@
+package com.societegenerale.commons.plugin.gradle;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+import com.societegenerale.commons.plugin.model.Rules;
+import com.societegenerale.commons.plugin.service.RuleInvokerService;
+import org.apache.commons.lang3.StringUtils;
+import org.gradle.api.GradleException;
+import org.gradle.workers.WorkAction;
+
+public abstract class WorkerAction implements WorkAction<WorkerActionParams> {
+
+    private static final String PREFIX_ARCH_VIOLATION_MESSAGE = "ArchUnit Gradle plugin reported architecture failures listed below :";
+
+    @Override
+    public void execute() {
+        RuleInvokerService ruleInvokerService = getParameters().getRuleInvokerService().get();
+        Rules rules = getParameters().getRules().get();
+
+        String ruleFailureMessage = "";
+
+        try {
+
+            ruleFailureMessage = ruleInvokerService.invokeRules(rules);
+
+            if (!StringUtils.isEmpty(ruleFailureMessage)) {
+                throw new GradleException(PREFIX_ARCH_VIOLATION_MESSAGE + " \n" + ruleFailureMessage);
+            }
+    /*
+            Class<?> c = Thread.currentThread().getContextClassLoader().loadClass(className);
+            Method greetingMethod = c.getMethod("getGreeting");
+            Object o = c.newInstance();
+            System.out.println(greetingMethod.invoke(o));
+     */
+
+        } catch ( InstantiationException | IllegalAccessException | InvocationTargetException e) {
+            e.printStackTrace();
+        }
+    }
+}

--- a/src/main/java/com/societegenerale/commons/plugin/gradle/WorkerAction.java
+++ b/src/main/java/com/societegenerale/commons/plugin/gradle/WorkerAction.java
@@ -1,13 +1,12 @@
 package com.societegenerale.commons.plugin.gradle;
 
 import java.lang.reflect.InvocationTargetException;
-import java.lang.reflect.Method;
 
-import com.societegenerale.commons.plugin.model.Rules;
 import com.societegenerale.commons.plugin.service.RuleInvokerService;
 import org.apache.commons.lang3.StringUtils;
 import org.gradle.api.GradleException;
 import org.gradle.workers.WorkAction;
+import org.slf4j.LoggerFactory;
 
 public abstract class WorkerAction implements WorkAction<WorkerActionParams> {
 
@@ -15,24 +14,22 @@ public abstract class WorkerAction implements WorkAction<WorkerActionParams> {
 
     @Override
     public void execute() {
-        RuleInvokerService ruleInvokerService = getParameters().getRuleInvokerService().get();
-        Rules rules = getParameters().getRules().get();
+
+        ArchUnitGradleConfig archUnitGradleConfig = getParameters().getArchUnitGradleConfig().get();
+
+        RuleInvokerService ruleInvokerService = new RuleInvokerService(new GradleLogAdapter(LoggerFactory.getLogger(RuleInvokerService.class)),
+                new GradleScopePathProvider(archUnitGradleConfig),
+                archUnitGradleConfig.getExcludedPaths());
 
         String ruleFailureMessage = "";
 
         try {
 
-            ruleFailureMessage = ruleInvokerService.invokeRules(rules);
+            ruleFailureMessage = ruleInvokerService.invokeRules(archUnitGradleConfig.getRules());
 
             if (!StringUtils.isEmpty(ruleFailureMessage)) {
                 throw new GradleException(PREFIX_ARCH_VIOLATION_MESSAGE + " \n" + ruleFailureMessage);
             }
-    /*
-            Class<?> c = Thread.currentThread().getContextClassLoader().loadClass(className);
-            Method greetingMethod = c.getMethod("getGreeting");
-            Object o = c.newInstance();
-            System.out.println(greetingMethod.invoke(o));
-     */
 
         } catch ( InstantiationException | IllegalAccessException | InvocationTargetException e) {
             e.printStackTrace();

--- a/src/main/java/com/societegenerale/commons/plugin/gradle/WorkerActionParams.java
+++ b/src/main/java/com/societegenerale/commons/plugin/gradle/WorkerActionParams.java
@@ -1,0 +1,13 @@
+package com.societegenerale.commons.plugin.gradle;
+
+import com.societegenerale.commons.plugin.model.Rules;
+import com.societegenerale.commons.plugin.service.RuleInvokerService;
+import org.gradle.api.provider.Property;
+import org.gradle.workers.WorkParameters;
+
+public interface WorkerActionParams extends WorkParameters {
+    Property<String> getClassName();
+
+    Property<RuleInvokerService> getRuleInvokerService();
+    Property<Rules> getRules();
+}

--- a/src/main/java/com/societegenerale/commons/plugin/gradle/WorkerActionParams.java
+++ b/src/main/java/com/societegenerale/commons/plugin/gradle/WorkerActionParams.java
@@ -1,6 +1,7 @@
 package com.societegenerale.commons.plugin.gradle;
 
-import com.societegenerale.commons.plugin.model.ConfigurableRule;
+import java.io.Serializable;
+import java.util.ArrayList;
 import org.gradle.api.provider.ListProperty;
 import org.gradle.api.provider.Property;
 import org.gradle.workers.WorkParameters;
@@ -16,6 +17,20 @@ public interface WorkerActionParams extends WorkParameters {
 
     ListProperty<String> getPreConfiguredRules();
 
-    ListProperty<ConfigurableRule> getConfigurableRules();
+    ListProperty<WorkerActionParams.ConfigurableRule> getConfigurableRules();
+
+    class ConfigurableRule implements Serializable {
+        static final long serialVersionUID = 1L;
+        String rule;
+        ApplyOn applyOn;
+        ArrayList<String> checks = new ArrayList<>();
+        boolean skip;
+    }
+
+    class ApplyOn implements Serializable {
+        static final long serialVersionUID = 1L;
+        String packageName;
+        String scope;
+    }
 
 }

--- a/src/main/java/com/societegenerale/commons/plugin/gradle/WorkerActionParams.java
+++ b/src/main/java/com/societegenerale/commons/plugin/gradle/WorkerActionParams.java
@@ -6,8 +6,6 @@ import org.gradle.api.provider.Property;
 import org.gradle.workers.WorkParameters;
 
 public interface WorkerActionParams extends WorkParameters {
-    Property<String> getClassName();
-
-    Property<RuleInvokerService> getRuleInvokerService();
+     Property<RuleInvokerService> getRuleInvokerService();
     Property<Rules> getRules();
 }

--- a/src/main/java/com/societegenerale/commons/plugin/gradle/WorkerActionParams.java
+++ b/src/main/java/com/societegenerale/commons/plugin/gradle/WorkerActionParams.java
@@ -1,11 +1,8 @@
 package com.societegenerale.commons.plugin.gradle;
 
-import com.societegenerale.commons.plugin.model.Rules;
-import com.societegenerale.commons.plugin.service.RuleInvokerService;
 import org.gradle.api.provider.Property;
 import org.gradle.workers.WorkParameters;
 
 public interface WorkerActionParams extends WorkParameters {
-     Property<RuleInvokerService> getRuleInvokerService();
-    Property<Rules> getRules();
+    Property<ArchUnitGradleConfig> getArchUnitGradleConfig();
 }

--- a/src/main/java/com/societegenerale/commons/plugin/gradle/WorkerActionParams.java
+++ b/src/main/java/com/societegenerale/commons/plugin/gradle/WorkerActionParams.java
@@ -1,16 +1,16 @@
 package com.societegenerale.commons.plugin.gradle;
 
+import com.societegenerale.commons.plugin.model.ConfigurableRule;
 import org.gradle.api.provider.ListProperty;
 import org.gradle.api.provider.Property;
 import org.gradle.workers.WorkParameters;
-
-import java.io.Serializable;
-import java.util.ArrayList;
 
 public interface WorkerActionParams extends WorkParameters {
     Property<String> getMainClassesPath();
 
     Property<String> getTestClassesPath();
+
+    Property<String> getProjectBuildDirPath();
 
     ListProperty<String> getExcludedPaths();
 
@@ -18,17 +18,4 @@ public interface WorkerActionParams extends WorkParameters {
 
     ListProperty<ConfigurableRule> getConfigurableRules();
 
-    class ConfigurableRule implements Serializable {
-        static final long serialVersionUID = 1L;
-        String rule;
-        ApplyOn applyOn;
-        ArrayList<String> checks = new ArrayList<>();
-        boolean skip;
-    }
-
-    class ApplyOn implements Serializable {
-        static final long serialVersionUID = 1L;
-        String packageName;
-        String scope;
-    }
 }

--- a/src/main/java/com/societegenerale/commons/plugin/gradle/WorkerActionParams.java
+++ b/src/main/java/com/societegenerale/commons/plugin/gradle/WorkerActionParams.java
@@ -1,8 +1,34 @@
 package com.societegenerale.commons.plugin.gradle;
 
+import org.gradle.api.provider.ListProperty;
 import org.gradle.api.provider.Property;
 import org.gradle.workers.WorkParameters;
 
+import java.io.Serializable;
+import java.util.ArrayList;
+
 public interface WorkerActionParams extends WorkParameters {
-    Property<ArchUnitGradleConfig> getArchUnitGradleConfig();
+    Property<String> getMainClassesPath();
+
+    Property<String> getTestClassesPath();
+
+    ListProperty<String> getExcludedPaths();
+
+    ListProperty<String> getPreConfiguredRules();
+
+    ListProperty<ConfigurableRule> getConfigurableRules();
+
+    class ConfigurableRule implements Serializable {
+        static final long serialVersionUID = 1L;
+        String rule;
+        ApplyOn applyOn;
+        ArrayList<String> checks = new ArrayList<>();
+        boolean skip;
+    }
+
+    class ApplyOn implements Serializable {
+        static final long serialVersionUID = 1L;
+        String packageName;
+        String scope;
+    }
 }

